keps/sig-api-machinery/NNNN-watch-list/README.md

```diff
@@ -0,0 +1,956 @@
+There are ways to do that cheaply, e.g. we could issue a count request against the datastore.
+Next, the cacher creates a new cacheWatcher (implements watch.Interface) passing the given bookmarkAfterResourceVersion, and gets initial data from the watchCache.
+After sending initial data the cacheWatcher starts listening on an incoming channel for new events, including a bookmark event.
+At some point, the cacher will receive an event with the resourceVersion equal or greater to the bookmarkAfterResourceVersion.
+It will be propagated to the cacheWatcher and then back to the reflector as a BOOKMARK event.
+
+![watch cache with the proposed changes](./cacher.png "watch cache with the proposed changes")
+
+Step 2a: Where does the initial data come from?
+
+During construction, the cacher creates the reflector and the watchCache.
+Since the watchCache implements the Store interface it is used by the reflector to store all data it has received from etcd.
+Note: Right now the reflector used to populate the watchCache uses the old semantics of LIST/WATCH.
+This could be optimized in the future (but out-of-scope here) by establishing a WATCH directly with etcd and waiting for the bookmarkAfterResourceVersion.
+
+Step 2b: What happens when new events are received while the cacheWatcher is sending initial data?
+
+The cacher maintains a list of all current watchers (cacheWatcher) and a separate goroutine (dispatchEvents) for delivering new events to the watchers.
+New events are added via the cacheWatcher.nonblockingAdd method that adds an event to the cacheWatcher.incoming channel.
+The cacheWatcher.incoming is a buffered channel and has a different size for different Resources (10 or 1000).
+Since the cacheWatcher starts processing the cacheWatcher.incoming channel only after sending all initial events it might block once its buffered channel tips over.
+In that case, it will be added to the list of blockedWatchers and will be given another chance to deliver an event after all nonblocking watchers have sent the event.
+All watchers that have failed to deliver the event will be closed.
+
+Note: Given the above, there is a possibility of sending initial data taking too long and failing the given time budget. 
+In that case, the cacheWatcher will be closed and the client (reflector) would have to reconnect.
+
+Note 2: Currently, there is a timeout for LIST requests of 60s.
+That means a slow reflector might fail synchronization as well and would have to re-establish the connection.
+
+Step 2c: How bookmarks are delivered to the cacheWatcher?
+
+First of all, the primary purpose of bookmark events is to deliver the current resourceVersion to watchers, continuously even without regular events happening.
+There are two sources of resourceVersions.
+The first one is regular events that contain RVs besides objects.
+The second one is a special type of etcd event called progressNotification delivering the most up-to-date revision with the given interval.
```

Comment:
those aren't delivered to end client at all - this only consumed only by kube-apiserver and not forwarded further

Reply:
clarified